programs/gamma/src/fees/dynamic_fee.rs

```diff
@@ -199,113 +217,90 @@ impl DynamicFee {
         Ok(dynamic_fee.min(MAX_FEE_VOLATILITY as u128) as u64)
     }
 
-    /// Gets the price range within a specified time window
+    /// Gets the price range within a specified time window and computes TWAP
     ///
     /// # Arguments
     /// * `observation_state` - Historical price observations
     /// * `current_time` - The current timestamp
     /// * `window` - The time window to consider
     ///
     /// # Returns
-    /// A tuple of (min_price, max_price, average_price) observed within the window
+    /// A tuple of (min_price, max_price, twap_price) observed within the window
     fn get_price_range(
         observation_state: &ObservationState,
         current_time: u64,
         window: u64,
     ) -> Result<(u128, u128, u128)> {
         let mut min_price = u128::MAX;
         let mut max_price = 0u128;
-        let mut total_price = 0u128;
+        let mut weighted_price_sum = Decimal::new(0, 0);
+        let mut total_weight = Decimal::new(0, 0);
 
-        let mut descending_order_observations = observation_state
+        // Collect valid observations within the window
+        let observations = observation_state
             .observations
             .iter()
-            .enumerate()
-            .filter(|(_, observation)| {
-                observation.block_timestamp != 0
-                    && observation.cumulative_token_0_price_x32 != 0
-                    && observation.cumulative_token_1_price_x32 != 0
-            })
-            .map(|(index, observation)| ObservationWithIndex {
-                index: index as u16,
-                observation: *observation,
+            .filter(|obs| {
+                obs.block_timestamp != 0
+                    && obs.cumulative_token_0_price_x32 != 0
+                    && current_time.saturating_sub(obs.block_timestamp) <= window
             })
             .collect::<Vec<_>>();
 
-        descending_order_observations.sort_by(|a, b| {
-            { b.observation.block_timestamp }.cmp(&{ a.observation.block_timestamp })
-        });
+        if observations.len() < 2 {
+            // Not enough data points to compute TWAP
+            return Ok((0, 0, 0));
+        }
 
-        let mut count = 0;
-        for observation_with_index in descending_order_observations {
-            let is_in_observation_window = current_time
-                .saturating_sub(observation_with_index.observation.block_timestamp)
-                <= window;
+        // Iterate over observation pairs to compute TWAP
+        for i in 0..observations.len() - 1 {
+            let obs = observations[i];
+            let next_obs = observations[i + 1];
```

Comment:
What about the case when the next observation is older than the current one?

This is possible as we have circular indexing.